tla-bmcmt/src/main/scala/at/forsyte/apalache/tla/bmcmt/rules/SetAsFunRule.scala

```diff
@@ -0,0 +1,80 @@
+package at.forsyte.apalache.tla.bmcmt.rules
+
+import at.forsyte.apalache.tla.bmcmt._
+import at.forsyte.apalache.tla.bmcmt.types.CellT
+import at.forsyte.apalache.tla.lir.UntypedPredefs._
+import at.forsyte.apalache.tla.lir._
+import at.forsyte.apalache.tla.lir.convenience.tla
+import at.forsyte.apalache.tla.lir.oper.ApalacheOper
+
+/**
+ * Implements a rule for Apalache!SetAsFun.
+ *
+ * @author Igor Konnov
+ */
+class SetAsFunRule(rewriter: SymbStateRewriter) extends RewritingRule {
+  override def isApplicable(symbState: SymbState): Boolean = {
+    symbState.ex match {
+      case OperEx(ApalacheOper.setAsFun, _) => true
+      case _                                => false
+    }
+  }
+
+  override def apply(state: SymbState): SymbState = {
+    state.ex match {
+      case OperEx(ApalacheOper.setAsFun, setEx) =>
+        var nextState = rewriter.rewriteUntilDone(state.setRex(setEx))
+        val setCell = nextState.asCell
+        setEx.typeTag match {
+          case Typed(SetT1(TupT1(keyType, valueType))) =>
+            nextState = translateRelation(setCell, nextState)
+            val rel = nextState.asCell
+            // construct a cell for the function and attach the relation to it
+            nextState = nextState.updateArena(_.appendCell(CellT.fromType1(FunT1(keyType, valueType))))
+            val fun = nextState.arena.topCell
+            nextState = nextState.updateArena(_.setCdm(fun, rel))
+            nextState.setRex(fun.toNameEx)
+
+          case tt =>
+            throw new RewriterException("Unexpected type in SetAsFunRule: " + tt, state.ex)
+        }
+
+      case _ =>
+        throw new RewriterException("%s is not applicable".format(getClass.getSimpleName), state.ex)
+    }
+  }
+
+  private def translateRelation(pairsSet: ArenaCell, state: SymbState): SymbState = {
```

Comment:
Please add a comment on the purpose of this method and the expected output (maybe a small example).